### DISCUSSION TITLE
Week 9 Checkin

### DISCUSSION THREAD

**Discussion by NaNa7Miiii** [UW-Upwelling-Project/Summer-2023#14] [General]
#

**Comment by NaNa7Miiii** [comment]
### Testing the Upwelling Classification Model
Since the model is trained using data from 2000 to 2020, my testings use data from 1979 to 1999:
<img width="559" alt="Screenshot 2023-08-15 at 4 12 08 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/d7160ede-c017-4a18-86a7-b25ca0001ee7">
<img width="850" alt="Screenshot 2023-08-15 at 4 12 33 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/38645dc7-8bb5-4143-ab10-871f1df7893b">
<img width="595" alt="Screenshot 2023-08-15 at 4 12 50 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/f5beefaf-9d1d-49ea-85c8-06906dc3da02">
I also want to interpret the model, and the package I am using is `lime`, I have a chunk for me to do visualization:
```
def preprocess_and_predict(image_array, model, target_shape=(149, 181)):
    assert image_array.shape == target_shape, "Data shape mismatch!"
    
    mean_val = np.nanmean(image_array)
    normalized_image = image_array - mean_val
    normalized_image = np.nan_to_num(normalized_image)
    preprocessed_image = normalized_image[np.newaxis, :, :, np.newaxis] 

    prediction = model.predict(preprocessed_image)
    return prediction[0][0]

explainer = lime_image.LimeImageExplainer()

segmenter = SegmentationAlgorithm('slic', n_segments=200, compactness=10, max_iter=10)

def model_predict(images_list):
    results = []
    for img in images_list:
        grayscale_img = img[:, :, 0]
        pred = preprocess_and_predict(grayscale_img, model)
        results.append([1-pred, pred])
    return np.array(results)

def plot_sample_with_mask(sample, mask, nan_mask, title=''):
    sample_2d = np.squeeze(sample).copy()
    sample_2d[nan_mask] = np.nan  # Restore NaN values for visualization
    
    plt.imshow(sample_2d, cmap='viridis')
    plt.imshow(mask, alpha=0.5, cmap='RdBu_r')
    
    plt.title(title)
    plt.colorbar()
    plt.show()

npy_path = "/home/jovyan/shared/jiarui-notebooks/2018-07-15.npy"
image_array = np.load(npy_path)
nan_mask = np.isnan(image_array)
target_shape = (149, 181)

explainer_image = image_array.copy()  
explainer_image = np.nan_to_num(explainer_image - np.nanmean(explainer_image))  # Normalize and remove NaN
explainer_image = explainer_image[:, :, np.newaxis]

explanation = explainer.explain_instance(explainer_image, model_predict, top_labels=5, hide_color=0, num_samples=3000, segmentation_fn=segmenter)

temp, mask = explanation.get_image_and_mask(1, positive_only=True, num_features=10, hide_rest=False)
plot_sample_with_mask(explainer_image, mask, nan_mask, title='LIME Explanation Overlay')
```
Here is the result that contains 10 features that the model considers the most important features to detect upwelling.
<img width="547" alt="Screenshot 2023-08-15 at 4 14 11 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/8e5665d1-beb6-4f46-89da-dc8dce25e755">
One can try generating more samples or using a different segmented to generate different highlighted areas. 

For the next step, I am thinking of building an Android APP that supports users to detect coastal upwelling from the SST image in their local drive, so I converted my model to .tflite format (This is one of the formats can be processed in Android Studio, another format that I know is .onnx) using:
```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

with open('upwelling_detection.tflite', 'wb') as f:
    f.write(tflite_model)
```

One can also try to build an ios APP, and the advantage is that coreml models are easy to deploy and can use ANE to accelerate prediction (low energy consumption on mobile devices)

**Comment by NaNa7Miiii** [comment]
Jiarui Tuesday Checkin:
**Trained a model with both air temperature and SST as input**: (data from 2010-2022)
<img width="784" alt="Screenshot 2023-08-16 at 3 50 06 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/1c62e913-fb02-4764-846f-3218fd1d632f">
<img width="757" alt="Screenshot 2023-08-16 at 3 50 24 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/359fc8aa-c7bf-4fe2-9f5a-fdf51f38583a">
**Trained a model with SST, air temperature, wind direction, wind speed as input:** (data from 2010-2022)
<img width="867" alt="Screenshot 2023-08-16 at 10 19 53 AM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/e89a2340-1c50-46df-aee3-1f3792913f38">
The accuracy does not increase much, and you can tell an overfitting occurs here. I plan to load the model and train it with data from 1979-1999.

**Comment by eeholmes** [comment]
# Data updates

itcoocean.2i2c.cloud upload

s3 buckets

From Luis Lopez: change bucket policy to
```
{
    "Version": "2012-10-17",
    "Id": "read-allow-policy",
    "Statement": [
        {
            "Sid": "allow-reads",
            "Effect": "Allow",
            "Principal": {
                "AWS": "*"
            },
            "Action": [
                "s3:GetObject",
                "s3:PutObject",
                "s3:GetBucketLocation",
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::YOUR_BUCKET/*",
                "arn:aws:s3:::YOUR_BUCKET"
            ],
            "Condition": {
                "StringEquals": {
                    "aws:RequestedRegion": "us-west-2"
                }
            }
        }
    ]
}
```

**Comment by minhphan03** [comment]
# Minh's Thursday Checkin:
- Set-up and configured AWS account & S3 buckets 
- Developed basic script to upload files directly from jhub using boto3

**Comment by NaNa7Miiii** [comment]
Jiarui Thursday Checkin:
Trained two models (variables are hold same as yesterday), both using today's data and tomorrow's label:
**SST & Air Temperature**:
<img width="888" alt="Screenshot 2023-08-17 at 2 06 45 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/7ffde8bc-3b22-49e5-a882-52614c02912c">
**Result: Test Accuracy is around 98%, with no sign of overfitting.** 

**SST & Air Temperature & Wind Speed & Wind Direction**:
<img width="888" alt="Screenshot 2023-08-17 at 2 07 43 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/92b09e76-658b-40de-912c-4d7c72080399">
**Result: Test Accuracy is around 98.45%, with no obvious overfitting.**

**Comment by NaNa7Miiii** [comment]
Jiarui Friday Checkin:
Making Streamlit web using the SST forecast model made in OHW
